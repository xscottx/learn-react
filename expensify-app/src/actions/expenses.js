import uuid from 'uuid';
import database from '../firebase/firebase';

// Action creators creation

// Pure react-redux
// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// Middleware style
// components calls action generator
// action generator returns function
// component dispatches function
// function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  })

export const startAddExpense = (expenseData = {}) => {
  // this returns a function, have access to getState from thunk
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0
    } = expenseData;

    const expense = {description, note, amount, createdAt};

    // this is a function that returns a promise
    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

// START_REMOVE_EXPENSE
export const startRemoveExpense = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    })
  }
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// START_EDIT_EXPENSE
export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    })
  }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// START_SET_EXPENSES
export const startSetExpenses = () => {
  // this returns a function
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    // this is a function that returns a promise
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
        const expenses = [];
      
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        });
      
        dispatch(setExpenses(expenses));
      })
  }
}