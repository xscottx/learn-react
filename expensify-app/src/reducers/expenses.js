// Reducers creation
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      // spread operator used to return a new collection without modifying existing state, this appends expense to array
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      // filter out object array inside expenses that do not match the expense id
      return state.filter(({id}) => id !== action.id);
    case 'EDIT_EXPENSE':
      // map will allow you to traverse thru array, and pass in an 'updater' function to make changes if needed
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }
        }
      })
    default:
      return state;
  }
}