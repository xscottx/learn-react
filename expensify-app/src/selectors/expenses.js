import moment from 'moment';

// Get visible expenses
// JS toLowerCase() lowers strings, and .includes() checks if the text is inside string
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {

  return expenses.filter((expense) => {
    // create a moment for expense and filter on dates before or after the start/end date match.
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    }
    else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
    
  })
}

export default getVisibleExpenses;