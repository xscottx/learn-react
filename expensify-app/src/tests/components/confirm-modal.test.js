import React from 'react';
import { shallow } from 'enzyme';
import ConfirmModal from '../../components/confirm-modal';
import expenses from '../fixtures/expenses';

let handleCancel, handleRemoveOption, wrapper;

beforeEach(() => {
  handleCancel = jest.fn(); // spy
  handleRemoveOption = jest.fn(); //spy
  wrapper = shallow(
    <ConfirmModal 
      selectedOption={expenses[0].description}
      handleCancel={handleCancel}
      handleRemoveOption={handleRemoveOption}
    />);
})

test('should render confirm modal', () => {
  expect(wrapper).toMatchSnapshot();  
})

test('should handle start remove expense', () => {
  wrapper.find('button').simulate('click');
  expect(handleRemoveOption).toHaveBeenCalled();
})