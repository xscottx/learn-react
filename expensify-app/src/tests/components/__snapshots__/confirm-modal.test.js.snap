// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render confirm modal 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Are you sure?"
  isOpen={true}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3
    className="modal__title"
  >
    Delete expense?
  </h3>
  <p
    className="modal__body"
  >
    Gum
  </p>
  <button
    className="button button--secondary"
    onClick={[MockFunction]}
  >
    Delete
  </button>
</Modal>
`;
